### PR TITLE
Add PodDisruptionBudget support in pod preemption

### LINKED ISSUE #53913: Support PodDisruptionBudget during preemption
**Is this a BUG REPORT or FEATURE REQUEST?**: FEATURE REQUEST

The alpha version of preemption in 1.8 does not support PodDisruptionBudget. Supporting PodDisruptionBudget is needed before we can move the feature to Beta.

/kind feature
/sig scheduling

### PR DESCRIPTION
**Pull request by bsalamat**
**What this PR does / why we need it**:
This PR adds the logic to make scheduler preemption aware of PodDisruptionBudget. Preemption tries to avoid preempting pods whose PDBs are violated by preemption. If preemption does not find any other pods to preempt, it will preempt pods despite violating their PDBs.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #53913

**Special notes for your reviewer**:

**Release note**:

```release-note
Add PodDisruptionBudget support during pod preemption
```

ref/ #47604

/sig scheduling

### CODE REVIEW

**Comment by bsalamat** [comment]
@davidopp This is my PR to add support for PDB. Could you please add milestone?
I am going to add integration tests.

File: plugin/pkg/scheduler/core/generic_scheduler.go [hunk 1]
```diff
@@ -625,51 +637,66 @@ func EqualPriorityMap(_ *v1.Pod, _ interface{}, nodeInfo *schedulercache.NodeInf
...
 			// We found a node that doesn't need any preemption. Return it!
 			// This should happen rarely when one or more pods are terminated between
 			// the time that scheduler tries to schedule the pod and the time that
 			// preemption logic tries to find nodes for preemption.
 			return node
 		}
+		numPDBViolatingPods := victims.numPDBViolations
+		if numPDBViolatingPods < minNumPDBViolatingPods {
+			minNumPDBViolatingPods = numPDBViolatingPods
+			minPDBViolatingNodes = nil
```
Comment (davidopp) [inline plugin/pkg/scheduler/core/generic_scheduler.go]: shouldn't this be `minPDBViolatingNodes = node` ?

Comment (davidopp) [reply plugin/pkg/scheduler/core/generic_scheduler.go]: oh never mind

File: plugin/pkg/scheduler/core/generic_scheduler.go [hunk 2]
```diff
@@ -776,6 +809,45 @@ func nodePassesExtendersForPreemption(
 	return true, nil
 }
 
+// filterPodsWithPDBViolation groups the given "pods" into two groups of "violatedPos"
```
Comment (davidopp) [inline plugin/pkg/scheduler/core/generic_scheduler.go]: violatedPods

File: plugin/pkg/scheduler/core/generic_scheduler.go [hunk 3]
```diff
@@ -788,14 +860,14 @@ func nodePassesExtendersForPreemption(
...
-// TODO(bsalamat): Add support for PodDisruptionBudget.
 func selectVictimsOnNode(
 	pod *v1.Pod,
 	meta algorithm.PredicateMetadata,
 	nodeInfo *schedulercache.NodeInfo,
 	fitPredicates map[string]algorithm.FitPredicate,
 	queue SchedulingQueue,
-) ([]*v1.Pod, bool) {
+	pdbs []*policy.PodDisruptionBudget,
+) ([]*v1.Pod, int, bool) {
```
Comment (davidopp) [inline plugin/pkg/scheduler/core/generic_scheduler.go]: Update the function-level comment to explain how you use PDB.

File: plugin/pkg/scheduler/core/generic_scheduler.go (hunk 2 above)
Comment (bsalamat) [reply plugin/pkg/scheduler/core/generic_scheduler.go]: Done

File: plugin/pkg/scheduler/core/generic_scheduler.go (hunk 3 above)
Comment (bsalamat) [reply plugin/pkg/scheduler/core/generic_scheduler.go]: Done

**Review by bsalamat — COMMENTED** [review]
Thanks, @davidopp! Addressed the comments and added more integration tests. No other logic changes. PTAL.

File: test/integration/scheduler/preemption_test.go [hunk 4]
```diff
@@ -495,3 +498,272 @@ func TestNominatedNodeCleanUp(t *testing.T) {
...
+					Namespace: context.ns.Name,
+					Priority:  &lowPriority,
+					Resources: defaultPodRes,
+					NodeName:  "node-2",
+					Labels:    map[string]string{"foo2": "bar"},
+				}),
+				initPausePod(context.clientSet, &pausePodConfig{
+					Name:      "low-pod3",
+					Namespace: context.ns.Name,
+					Priority:  &lowPriority,
```
Comment (davidopp) [inline test/integration/scheduler/preemption_test.go]: I'd suggest to make this $mediumPriority so it exactly matches node-1 except for violating more PDBs

**Comment by davidopp** [comment]
Looks great, just had one small comment.

File: test/integration/scheduler/preemption_test.go (hunk 4 above)
Comment (bsalamat) [reply test/integration/scheduler/preemption_test.go]: Done.

**Comment by bsalamat** [comment]
Thanks a lot, @davidopp for your prompt review. PTAL.

**Comment by davidopp** [comment]
PR is LGTM and approved, just awaiting auto-merge.

**Comment by bsalamat** [comment]
I wonder why this has not auto-merged yet.

**Pull request merged**